src/tools/fuzzing/heap-types.cpp

```diff
@@ -0,0 +1,492 @@
+#include "tools/fuzzing/parameters.h"
+
+namespace wasm {
+
+namespace {
+
+struct HeapTypeGenerator {
+  Random& rand;
+  FeatureSet features;
+  TypeBuilder builder;
+
+  // Map the HeapTypes we are building to their indices in the builder.
+  std::unordered_map<HeapType, Index> typeIndices;
+
+  // For each type we will build, the indices of its subtypes we will build.
+  std::vector<std::vector<Index>> subtypeIndices;
+
+  // Abstract over all the types that may be assigned to a builder slot.
+  using Assignable =
+    std::variant<HeapType::BasicHeapType, Signature, Struct, Array>;
+
+  // Top-level kinds, chosen before the types are actually constructed. This
+  // allows us to choose HeapTypes that we know will be subtypes of data or func
+  // before we actually generate the types.
+  using BasicKind = HeapType::BasicHeapType;
+  struct SignatureKind {};
+  struct DataKind {};
+  using HeapTypeKind = std::variant<BasicKind, SignatureKind, DataKind>;
+  std::vector<HeapTypeKind> typeKinds;
+
+  HeapTypeGenerator(Random& rand, FeatureSet features, size_t n)
+    : rand(rand), features(features), builder(n), subtypeIndices(n) {}
+
+  HeapType::BasicHeapType generateBasicHeapType() {
+    return rand.pick(HeapType::func,
+                     HeapType::ext,
+                     HeapType::any,
+                     HeapType::eq,
+                     HeapType::i31,
+                     HeapType::data);
+  }
+
+  Type::BasicType generateBasicType() {
+    // Exclude basic reference types because we will construct them via the
+    // TypeBuilder.
+    return rand.pick(
+      Random::FeatureOptions<Type::BasicType>{}
+        .add(FeatureSet::MVP, Type::i32, Type::i64, Type::f32, Type::f64)
+        .add(FeatureSet::SIMD, Type::v128));
+  }
+
+  HeapType generateHeapType() {
+    if (rand.oneIn(2)) {
+      return generateBasicHeapType();
+    } else {
+      return builder[rand.upTo(builder.size())];
+    }
+  }
+
+  // TODO: Create an enum for `Defaultable, NonDefaultable` in wasm-type.h.
+  Type generateRefType(bool defaultable = false) {
+    auto heapType = generateHeapType();
+    auto nullability = (defaultable || rand.oneIn(2)) ? Nullable : NonNullable;
+    return builder.getTempRefType(heapType, nullability);
+  }
+
+  Type generateRttType() {
+    auto heapType = generateHeapType();
+    auto depth = rand.oneIn(2) ? Rtt::NoDepth : rand.upTo(MAX_RTT_DEPTH);
+    return builder.getTempRttType(Rtt(depth, heapType));
+  }
+
+  Type generateSingleType(bool defaultable = false) {
+    switch (rand.upTo(defaultable ? 2 : 3)) {
+      case 0:
+        return generateBasicType();
+      case 1:
+        return generateRefType(defaultable);
+      case 2:
+        return generateRttType();
+    }
+    WASM_UNREACHABLE("unexpected");
+  }
+
+  Type generateTupleType() {
+    std::vector<Type> types(2 + rand.upTo(MAX_TUPLE_SIZE - 1));
+    for (auto& type : types) {
+      // Make sure tuples are defaultable. See comment in getTupleType.
```

Comment:
This is no `getTupleType` function in this file..?

Reply:
Oops, this must be stale or copied from elsewhere incorrectly. Thanks for the catch.